nixos/modules/services/security/vault-agent.nix

```diff
@@ -0,0 +1,145 @@
+{config, lib, pkgs, ...}:
+
+with lib;
+
+let
+  cfg = config.services.vault-agent;
+  createAgentInstance = cfg: name:
+    let
+      configFile = pkgs.writeText "${name}.hcl" ''
+          pid_file = "/run/vault-agent/${name}.pid"
+          vault {
+            address = "${cfg.address}"
+          }
+          ${cfg.extraConfig}
+          ${mkTemplates cfg.templates}
+        '';
+      mkTemplates = templates:
+        concatStringsSep "\n"
+          (mapAttrsToList
+            (template: config: ''
+                template {
+                  source      = "${pkgs.writeText template config.content}"
+                  ${optionalString (config.command != null)
+                  "command     = \"${config.command}\""}
+                  destination = "${config.destination}"
+                  perms       = "${config.perms}"
+                }
+              '')
+              templates);
+    in  {
+      description = "Vault-agent ${name} daemon";
+      wantedBy = ["multi-user.target"];
+      after = [ "network.target" ];
+      startLimitIntervalSec = 60;
+      startLimitBurst = 3;
+      serviceConfig = {
+        User = "root";
+        Group = "root";
+        ExecStart = "${pkgs.vault}/bin/vault agent -config ${configFile}";
```

Comment:
`vault agent` and `consul-template` are incredibly similar to one another.
I think it would be great to somehow support both of them with a single module.

This could be done with just two more module options:
- `package` (`lib.mkPackageOption`) using `lib.getExe` instead of hard-coding `/bin/<binary`
- `type` or whatever, to insert that sub-comment (`vault agent` vs `consul-template`). Could also probably get auto-detected

---

Would also be cool to have a `extraFlags` option, which get passed to `ExecStart` :)